OpenRA.Mods.Cnc/Traits/World/TSEditorResourceLayer.cs

```diff
@@ -0,0 +1,110 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2021 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Cnc.Traits
+{
+	class TSEditorResourceLayerInfo : EditorResourceLayerInfo, Requires<EditorActorLayerInfo>
+	{
+		public readonly string VeinType = "Veins";
+
+		[ActorReference]
+		[Desc("Actor types that should be treated as veins for adjacency.")]
+		public readonly HashSet<string> VeinholeActors = new HashSet<string> { };
+
+		public override object Create(ActorInitializer init) { return new TSEditorResourceLayer(init.Self, this); }
+	}
+
+	class TSEditorResourceLayer : EditorResourceLayer
+	{
+		readonly TSEditorResourceLayerInfo info;
+		readonly EditorActorLayer actorLayer;
+
+		public TSEditorResourceLayer(Actor self, TSEditorResourceLayerInfo info)
+			: base(self, info)
+		{
+			this.info = info;
+			actorLayer = self.Trait<EditorActorLayer>();
+		}
+
+		bool IsValidVeinNeighbour(CPos cell, CPos neighbour)
+		{
+			// Cell is automatically valid if it contains a veinhole actor
+			if (actorLayer.PreviewsAt(neighbour).Any(a => info.VeinholeActors.Contains(a.Info.Name)))
+				return true;
+
+			// Neighbour must be flat or a cardinal slope, unless the resource cell itself is a slope
+			if (Map.Ramp[cell] == 0 && Map.Ramp[neighbour] > 4)
+				return false;
+
+			var terrainInfo = Map.Rules.TerrainInfo;
+			var terrainType = terrainInfo.TerrainTypes[terrainInfo.GetTerrainInfo(Map.Tiles[neighbour]).TerrainType].Type;
+			return info.ResourceTypes[info.VeinType].AllowedTerrainTypes.Contains(terrainType);
+		}
+
+		protected override bool AllowResourceAt(string resourceType, CPos cell)
+		{
+			var mapResources = Map.Resources;
+			if (!mapResources.Contains(cell))
+				return false;
+
+			// Resources are allowed on flat terrain and cardinal slopes
+			if (Map.Ramp[cell] > 4)
+				return false;
+
+			if (!info.ResourceTypes.TryGetValue(resourceType, out var resourceInfo))
+				return false;
+
+			// Ignore custom terrain types when spawning resources in the editor
+			var terrainInfo = Map.Rules.TerrainInfo;
+			var terrainType = terrainInfo.TerrainTypes[terrainInfo.GetTerrainInfo(Map.Tiles[cell]).TerrainType].Type;
+			if (!resourceInfo.AllowedTerrainTypes.Contains(terrainType))
+				return false;
+
+			// Veins must be placed next to a compatible border cell
+			if (resourceType == info.VeinType)
+			{
+				var neighboursValid = Common.Util.ExpandFootprint(cell, false)
+					.All(c => IsValidVeinNeighbour(cell, c));
```

Comment:
This is the bit that handles:

> The original TS engine implemented resources as overlays (along with walls, bridges, crates, etc), which meant that harvestable resources were automatically not allowed to overlap with visual only things. Veins were a bit weird in that the LAT border cells were not themselves harvestable (they gave a harvest cursor but the harvester would go to the closest interior cell), which means that our `ResourceLayer` code needs to reserve an empty cell between veins and tiberium so that the `ResourceRenderer` can draw the borders without it overlapping with tiberium.

Reply:
So because we need to have a visual-only border around the actual vein cells, all neighbours of vein cells must also be valid vein cells for the illusion to work? 🤔

Reply:
Non-diagonal neighbours, yeah.